File: C/4/1.c
#include <stdio.h>

/* Maximum input line length */
#define MAXLINE 1000

int getline(char line[], int max);
int strindex(char source[], char searchfor[]);

/* Pattern to be searched for. */
char pattern[] = "ould";

/**
 * A program to find all lines matching pattern.
 */
main()
{
    char line[MAXLINE];

    int pos = 0;

    while (getline(line, MAXLINE) > 0) {
        pos = strindex(line, pattern);
        printf("[*] Rightmost position of '%s' in '%s' is: '%d'\n", pattern, line, pos);
    }

    return 0;
}

/**
 * A function to read a line into s.
 *
 * @param char[] s      A string to save the read line at.
 * @param int    lim    Maximum length of a line.
 *
 * @return int Length of read string.
 */
int getline(char s[], int lim)
{
    int c, i;

    i = 0;
    while (--lim > 0 && (c=getchar()) != EOF && c != '\n')
        s[i++] = c;

    s[i] = '\0';

    return i;
}

/**
 * A function to check for the rightmost occurence of t in s.
 *
 * @param char[] s The string to be searched at
 * @param char[] t The pattern to be searched for
 *
 * @return int      Position of rightmost occurrence .
 */
int strindex(char s[], char t[])
{
    int i, j, k, rightmost = -1;

    for (i = 0; s[i] != '\0'; i++) {
        for (j=i, k=0; t[k]!='\0' && s[j]==t[k]; j++, k++);
        if (k > 0 && t[k] == '\0') {
            rightmost = i;
        }
    }

    return rightmost;
}

File: rustlang/4/1.rs
use std::io;

// Pattern to be searched for.
const PATTERN : &str = "ould";

/// A program to test pattern search.
fn main()
{
    let mut input = String::new();

    loop {
        match io::stdin().read_line(&mut input) {
            Ok (_bytes) => {
                if _bytes == 0 {
                    break;
                }

                let trimmed_input = input.trim_end();
                let pos: i32 = strindex(trimmed_input, PATTERN);

                print!("[*] Rightmost position of '{}' in '{}' is: '{}'\n", PATTERN, trimmed_input, pos);

                // clear input so lines don't get appended
                input.clear();
            },
            Err(_) => {
                panic!("Unexpected error.");
            }
        }
    }
}

/// A function to check for the rightmost occurrence of t in s.
///
/// @param &str s The string to be searched at
/// @param &str t The pattern to be searched for
///
/// @return i32   Position of rightmost occurrence of t inside s.
fn strindex(s: &str, t: &str) -> i32
{
    let mut rightmost : i32 = -1;

    for (idx, _) in s.chars().enumerate() {
        let mut str_iter = s.chars().skip(idx+1);
        let mut pattern_iter = t.chars();
        let mut k = 0;
        let pend;

        loop {
            let _sc = str_iter.next();
            let _pc = pattern_iter.next();

            if _pc == None || _sc != _pc {
                pend = _pc;
                break;
            }


            k += 1;
        };

        if k > 0 && pend == None {
            rightmost = (idx+1) as i32;
        }
    }

    rightmost
}